### PR TITLE
feat(rest): Implement rest module

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

discord.js-next is alive! Part 1: rest client should be working, but testing IS REQUIRED! If you're willing to test this, take care, any rate limit issues should be reported here, or on Discord

**Status**

- [ ] Code changes have been tested against the Discord API, or there are no code changes

**Semantic versioning classification:**

- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: packages/rest/src/lib/REST.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	 *
+	 * @default '8'
+	 */
+	version: string;
+}
+
+export class REST extends EventEmitter {
+	public readonly options: RESTOptions;
+	public readonly cdn: CDN;
+	public readonly requestManager: RequestManager;
```
Comment (Fyko) [inline packages/rest/src/lib/REST.ts:66]:
```suggestion
	public readonly requestManager = new RestManager(this);
```

File: packages/rest/src/lib/REST.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	public constructor(options: Partial<RESTOptions> = {}) {
+		super();
+		this.options = mergeDefault(DefaultRestOptions, options);
+		this.options.offset = Math.max(0, this.options.offset);
+
+		this.cdn = new CDN(this.options.cdn);
+		this.requestManager = new RequestManager(this);
```
Comment (Fyko) [inline packages/rest/src/lib/REST.ts:74]:
```suggestion
```

File: packages/rest/src/lib/RequestManager.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
+import { Agent } from 'https';
+import { Collection } from '@discordjs/collection';
+import type { REST } from './REST';
+import type { IHandler } from './handlers/IHandler';
+import { DiscordSnowflake } from '@sapphire/snowflake';
+import { SequentialHandler } from './handlers/SequentialHandler';
+import type { RequestInit } from 'node-fetch';
+import { DefaultUserAgent } from './utils/constants';
+import FormData from 'form-data';
```
Comment (Fyko) [inline packages/rest/src/lib/RequestManager.ts:9]: Should we use `@discordjs/form-data` for this?

File: packages/rest/src/lib/RequestManager.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth === undefined || request.auth) {
+			// If we haven't received a token, throw an error
+			if (!this.#token)
+				return Promise.reject(new Error('Expected token to be set for this request, but none was present'));
+
+			headers.Authorization = `Bot ${this.#token}`;
```
Comment (Fyko) [inline packages/rest/src/lib/RequestManager.ts:195]: Would it be a good idea to add an option for the token prefix so this can be used with access tokens?

File: packages/rest/src/lib/RequestManager.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+			// Strip out all IDs
+			.replace(/\d{16,19}/g, ':id')
+			// Strip out reaction as they fall under the same bucket
+			.replace(/\/reactions\/(.*)/, '/reactions/:reaction');
+
+		let exceptions = '';
+
+		// Hard-Code Old Message Deletion Exception (2 week+ old messages are a different bucket)
+		// https://github.com/discordapp/discord-api-docs/issues/1295
+		if (method === 'delete' && baseRoute === '/channels/:id/messages/:id') {
```
Comment (Fyko) [inline packages/rest/src/lib/RequestManager.ts:264]: Would it be beneficial to add a stricter check on the method here?  As a runtime check if a user decides to provide a string for their `REST#request` method instead of using the enum?

File: packages/rest/src/lib/CDN.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	extension?: ImageExtension;
+	size?: ImageSize;
+	dynamic?: boolean;
+}
+
+/**
+ * The CDN link builder
+ */
+export class CDN {
+	public constructor(private readonly base: string) {}
```
Comment (Fyko) [inline packages/rest/src/lib/CDN.ts:17]:
```suggestion
import { DefaultRestOptions } from './utils/constants';

const ALLOWED_EXTENSIONS = ['webp', 'png', 'jpg', 'jpeg', 'gif'] as const;
const ALLOWED_SIZES = [16, 32, 64, 128, 256, 512, 1024, 2048, 4096] as const;

export type ImageExtension = typeof ALLOWED_EXTENSIONS[number];
export type ImageSize = typeof ALLOWED_SIZES[number];

export interface ImageURLOptions {
	extension?: ImageExtension;
	size?: ImageSize;
	dynamic?: boolean;
}

/**
 * The CDN link builder
 */
export class CDN {
	public constructor(private readonly base: string = DefaultRestOptions.cdn) {}
```

**Review by Fyko — CHANGES_REQUESTED** [review]

File: packages/rest/src/lib/CDN.ts (hunk 6 above)
Comment (pizzafox) [reply packages/rest/src/lib/CDN.ts:17]: `ALLOWED_EXTENSIONS` and `ALLOWED_SIZES` should be `Set`s for faster validation (L135 and L139).

Comment (kyranet) [reply packages/rest/src/lib/CDN.ts:17]: They're not faster for small datasets, hashing a value takes a lot more time than iterating over 10 entries.

File: packages/rest/src/lib/RequestManager.ts (hunk 5 above)
Comment (vladfrangu) [reply packages/rest/src/lib/RequestManager.ts:264]: Unless we want to start validating everything you input in our module, with something like, say, [`ow`](https://npmjs.org/package/ow), if you're manually using `Rest#request` it means you know what you're doing.

File: packages/rest/src/lib/REST.ts (hunk 1 above)
Comment (zajrik) [reply packages/rest/src/lib/REST.ts:66]: Implicit typing is eww

File: packages/rest/src/lib/REST.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+export class REST extends EventEmitter {
+	public readonly options: RESTOptions;
+	public readonly cdn: CDN;
+	public readonly requestManager: RequestManager;
+
+	public constructor(options: Partial<RESTOptions> = {}) {
+		super();
+		this.options = mergeDefault(DefaultRestOptions, options);
+		this.options.offset = Math.max(0, this.options.offset);
```
Comment (kyranet) [inline packages/rest/src/lib/REST.ts:71]: We shouldn't mutate options, instead, use object spread.

File: packages/rest/src/lib/RequestManager.ts [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+			request.fullRoute
+		}${query}`;
+
+		// Create the required headers
+		const headers: RequestHeaders = {
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth === undefined || request.auth) {
```
Comment (kyranet) [inline packages/rest/src/lib/RequestManager.ts:195]: `request.auth` has 3 possible values: `undefined`, `false`, and `true`. This branch is run when it's either `undefined` or `true`, or in simpler terms: when it's not `false`.
```suggestion
		if (request.auth !== false) {
```

Comment (kyranet) [reply packages/rest/src/lib/RequestManager.ts:195]: Also, the headers from lines 189-192 and this `if` can be moved to nearly the top, that way `requst.query` and `url` aren't computed if a potential rejection (line 198) is caused.

File: packages/rest/src/lib/RequestManager.ts [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		this.handlers.set(queue.id, queue);
+
+		return queue;
+	}
+
+	/**
+	 * Formats the request data to a usable format for fetch
+	 * @param request The request data
+	 */
+	private resolveRequest(request: InternalRequest): Promise<{ url: string; fetchOptions: RequestInit }> {
```
Comment (kyranet) [inline packages/rest/src/lib/RequestManager.ts:176]: Why does this method return a Promise, when it's fully synchronous?

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/rest/src/lib/REST.ts (hunk 7 above)
Comment (vladfrangu) [reply packages/rest/src/lib/REST.ts:71]: Doesn't mergeDefault already cover this?

File: packages/rest/src/lib/RequestManager.ts (hunk 9 above)
Comment (vladfrangu) [reply packages/rest/src/lib/RequestManager.ts:176]: Beecauseeee... if we throw an error, it'd be a sync error, till we get to queueRequest on the handler, where it would be an async error... and otherwise we have to disable require-await in the original function

File: packages/rest/src/lib/REST.ts (hunk 7 above)
Comment (kyranet) [reply packages/rest/src/lib/REST.ts:71]: > Doesn't mergeDefault already cover this?

[Nope](https://github.com/sapphire-project/utilities/blob/01c440a2d5cff07effe22d64dc7929a9f9efa330/packages/utilities/src/lib/mergeDefault.ts#L46-L61).

Comment (Fyko) [reply packages/rest/src/lib/REST.ts:71]: @kyranet Unless I'm mistaken, `mergeDefault` depends on `deepClone` which [doesn't mutate](https://github.com/sapphire-project/utilities/blob/01c440a2d5cff07effe22d64dc7929a9f9efa330/packages/utilities/src/lib/deepClone.ts#L17)

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+			return await this.runRequest(routeID, url, options);
+		} finally {
+			// Allow the next request to fire
+			this.#asyncQueue.shift();
+		}
+	}
+
+	/**
+	 * The method that actually makes the request to the api, and updates info about the bucket accordingly
+	 * @param route The generalized api route with literal ids for major parameters
```
Comment (vaporox) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:119]: same as above

File: packages/rest/src/lib/RequestManager.ts [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+		// If a reason was set, set it's appropriate header
+		if (request.reason?.length) {
+			headers['X-Audit-Log-Reason'] = encodeURIComponent(request.reason);
+		}
+
+		// Format the full request URL (api base, optional version, endpoint, optional querystring)
+		const url = `${options.api}${request.versioned === false ? '' : `/v${options.version}`}${
+			request.fullRoute
+		}${query}`;
```
Comment (vaporox) [inline packages/rest/src/lib/RequestManager.ts:205]:
```suggestion
		const url = `${options.api}${request.versioned ? `/v${options.version}` : ''}${request.fullRoute}${query}`;
```

File: packages/rest/src/lib/errors/DiscordAPIError.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+	message: string;
+	errors?: DiscordError;
+}
+
+function isErrorGroupWrapper(error: any): error is DiscordErrorGroupWrapper {
+	return Reflect.has(error, '_errors');
+}
+
+function isErrorResponse(error: any): error is DiscordErrorFieldInformation {
+	return typeof Reflect.get(error, 'message') === 'string';
```
Comment (vaporox) [inline packages/rest/src/lib/errors/DiscordAPIError.ts:23]:
```suggestion
	return typeof error.message === 'string';
```

File: packages/rest/src/lib/RequestManager.ts [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+		// If a query option is passed, use it
+		if (request.query) query = `?${request.query.toString() as string}`;
+
+		// Create the required headers
+		const headers: RequestHeaders = {
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth !== false) {
```
Comment (vaporox) [inline packages/rest/src/lib/RequestManager.ts:190]:
```suggestion
		if (request.auth) {
```
not sure what values `request.auth` can have, but i think this is not needed

File: packages/rest/src/lib/RequestManager.ts [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+	 * Formats the request data to a usable format for fetch
+	 * @param request The request data
+	 */
+	private resolveRequest(request: InternalRequest): { url: string; fetchOptions: RequestInit } {
+		const { options } = this;
+
+		let query = '';
+
+		// If a query option is passed, use it
+		if (request.query) query = `?${request.query.toString() as string}`;
```
Comment (vaporox) [inline packages/rest/src/lib/RequestManager.ts:182]:
```suggestion
		// If a query option is passed, use it
		const query = request.query ? `?${request.query}` : '';
```

File: packages/rest/src/lib/RequestManager.ts [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+	public constructor(public readonly options: RESTOptions) {
+		super();
+	}
+
+	public setToken(token: string) {
+		this.#token = token;
+		return this;
+	}
+
+	public async queueRequest(request: InternalRequest): Promise<unknown> {
```
Comment (vaporox) [inline packages/rest/src/lib/RequestManager.ts:139]: these methods are missing documentation

File: packages/rest/src/lib/REST.ts [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+	public patch(fullRoute: string, options: RequestData = {}) {
+		return this.request({ ...options, fullRoute, method: RequestMethod.Patch });
+	}
+
+	/**
+	 * Runs a request from the api
+	 * @param options Request options
+	 */
+	public request(options: InternalRequest) {
+		return this.requestManager.queueRequest({ ...options });
```
Comment (vaporox) [inline packages/rest/src/lib/REST.ts:132]:
```suggestion
		return this.requestManager.queueRequest(options);
```

File: packages/rest/src/lib/REST.ts [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+	public constructor(options: Partial<RESTOptions> = {}) {
+		super();
+		this.options = mergeDefault(DefaultRestOptions, deepClone(options));
+		this.options.offset = Math.max(0, this.options.offset);
+
+		this.cdn = new CDN(this.options.cdn);
+		this.requestManager = new RequestManager(this.options);
+	}
+
+	public setToken(token: string) {
```
Comment (vaporox) [inline packages/rest/src/lib/REST.ts:77]: missing documentation

File: packages/rest/src/lib/utils/constants.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import { APIVersion } from 'discord-api-types/v8';
+import type { RESTOptions } from '../REST';
+// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
+const Package = require('../../../package.json');
```
Comment (vaporox) [inline packages/rest/src/lib/utils/constants.ts:4]: why not use `import` here?

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+	 * @param message The message to debug
+	 */
+	private debug(message: string) {
+		this.manager.emit(RESTEvents.Debug, `[REST ${this.id}] ${message}`);
+	}
+
+	/**
+	 * Queues a request to be sent
+	 * @param route The generalized api route with literal ids for major parameters
+	 * @param request All the information needed to make a request
```
Comment (vaporox) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:86]: same as above

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+	/**
+	 * The interface used to sequence async requests sequentially
+	 */
+	// eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+	#asyncQueue = new AsyncQueue();
+
+	/**
+	 * @param manager The rest manager
+	 * @param hash The hash that this RequestHandler handles
+	 * @param token The bot token used to make requests
```
Comment (vaporox) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:44]: docs need to be updated, parameters are different than the ones listed here

File: packages/rest/src/lib/REST.ts [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+import { InternalRequest, RequestData, RequestManager, RequestMethod } from './RequestManager';
+import { DefaultRestOptions } from './utils/constants';
+
+/**
+ * The events that the REST manager emits
+ */
+export const enum RESTEvents {
+	Debug = 'restDebug',
+	RateLimited = 'rateLimited',
+}
```
Comment (vaporox) [inline packages/rest/src/lib/REST.ts:13]: this could be moved to `constants.ts`

File: packages/rest/src/lib/CDN.ts [hunk 22]
```diff
@@ -0,0 +1,149 @@
+import { DefaultRestOptions } from './utils/constants';
+
+const ALLOWED_EXTENSIONS = ['webp', 'png', 'jpg', 'jpeg', 'gif'] as const;
+const ALLOWED_SIZES = [16, 32, 64, 128, 256, 512, 1024, 2048, 4096] as const;
```
Comment (vaporox) [inline packages/rest/src/lib/CDN.ts:4]: these could be moved to `constants.ts`

File: packages/rest/src/lib/REST.ts [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+
+/**
+ * Options to be passed when creating the REST instance
+ */
+export interface RESTOptions {
+	/**
+	 * The base api path, without version
+	 *
+	 * @default 'https://discord.com/api'
+	 */
```
Comment (vaporox) [inline packages/rest/src/lib/REST.ts:23]: i don't think there should be a newline here, this is rather inconsistent across the pr

File: packages/rest/src/lib/RequestManager.ts (hunk 11 above)
Comment (vladfrangu) [reply packages/rest/src/lib/RequestManager.ts:205]: N/A, done by our lint rules

File: packages/rest/src/lib/RequestManager.ts (hunk 13 above)
Comment (vladfrangu) [reply packages/rest/src/lib/RequestManager.ts:190]: It is needed, as it's meant to default to `true`, unless you manually set it to `false`

File: packages/rest/src/lib/errors/DiscordAPIError.ts (hunk 12 above)
Comment (vladfrangu) [reply packages/rest/src/lib/errors/DiscordAPIError.ts:23]: Technically, error should be typed as unknown, thus needing to keep the Reflect.get code. Will fix 👍

File: packages/rest/src/lib/RequestManager.ts (hunk 14 above)
Comment (vladfrangu) [reply packages/rest/src/lib/RequestManager.ts:182]: image

File: packages/rest/src/lib/RequestManager.ts (hunk 13 above)
Comment (vaporox) [reply packages/rest/src/lib/RequestManager.ts:196]: but as long as it's a boolean, those two should have the same result?

File: packages/rest/src/lib/RequestManager.ts (hunk 11 above)
Comment (kyranet) [reply packages/rest/src/lib/RequestManager.ts:211]: The explicit `=== false` is intentional, see https://github.com/discordjs/discord.js-next/pull/34#discussion_r565669214 for reference.

File: packages/rest/src/lib/RequestManager.ts (hunk 13 above)
Comment (kyranet) [reply packages/rest/src/lib/RequestManager.ts:196]: It can also be `undefined`, both `true` and `undefined` must trigger this branch.

File: packages/rest/src/lib/CDN.ts [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	/**
+	 * Generates a user avatar URL for a user's avatar.
+	 * @param userID The user ID that has the icon
+	 * @param avatarHash The hash provided by Discord for this avatar
+	 * @param options Optional options for the avatar
+	 */
+	public userAvatar(userID: string, avatarHash: string, { dynamic = false, ...options }: ImageURLOptions = {}): string {
+		if (dynamic) options.extension = avatarHash.startsWith('a_') ? 'gif' : options.extension;
```
Comment (NotSugden) [inline packages/rest/src/lib/CDN.ts:119]:
```suggestion
		if (dynamic && avatarHash.startsWith('a_')) options.extension = 'gif';
```
personally would go with this rather than the tenary

File: packages/rest/src/lib/REST.ts [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+	version: string;
+}
+
+export class REST extends EventEmitter {
+	public readonly cdn: CDN;
+	public readonly requestManager: RequestManager;
+
+	public constructor(options: Partial<RESTOptions> = {}) {
+		super();
+		this.cdn = new CDN(options.cdn ?? DefaultRestOptions.cdn);
```
Comment (kyranet) [inline packages/rest/src/lib/REST.ts:54]: The default is unnecessary as `CDN`'s constructor already defaults to this.
```suggestion
		this.cdn = new CDN(options.cdn);
```

**Review by kyranet — DISMISSED** [review]

File: packages/rest/src/lib/REST.ts (hunk 25 above)
Comment (vladfrangu) [reply packages/rest/src/lib/REST.ts:54]: Intended `??`
image

**Review by kyranet — APPROVED** [review]

File: packages/rest/package.json [hunk 26]
```diff
@@ -36,10 +39,21 @@
...
+		"discord-api-types": "^0.12.1",
+		"node-fetch": "^2.6.1",
+		"tslib": "^2.1.0"
 	},
-	"devDependencies": {},
 	"engines": {
 		"node": ">=14.0.0"
+	},
+	"devDependencies": {
+		"@types/nock": "^11.1.0",
```
Comment (Fyko) [inline packages/rest/package.json:56]: Types is installed but not the actual `nock` package?

**Review by Fyko — CHANGES_REQUESTED** [review]

File: packages/rest/package.json (hunk 26 above)
Comment (vladfrangu) [reply packages/rest/package.json:56]: Oops... Both nock and it's types need to be installed on root, not on this package. Will fix!

**Review by kyranet — DISMISSED** [review]

File: .github/workflows/lint-cron.yml [hunk 27]
```diff
@@ -10,13 +10,16 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v2
 
-      - name: Install Node v12
+      - name: Install Node v15
```
Comment (iCrawl) [inline .github/workflows/lint-cron.yml:13]: Can we upgrade this to 16? Just to be on an LTS version.

File: jest.config.js [hunk 28]
```diff
@@ -1,17 +1,24 @@
...
-			lines: 80,
-			statements: 80,
+			branches: 70,
+			lines: 70,
+			statements: 70,
 		},
 	},
 	roots: ['<rootDir>packages/'],
+	coveragePathIgnorePatterns: [
+		'/node_modules/', //
```
Comment (iCrawl) [inline jest.config.js:20]: We should just let prettier do it's thing here.

File: packages/rest/__tests__/DiscordAPIError.test.ts [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+		50035,
+		400,
+		'POST',
+		'https://discord.com/api/v8/channels/:id',
+	);
+
+	expect(error.code).toBe(50035);
+	expect(error.message).toBe(
+		[
+			'Invalid Form Body', //
```
Comment (iCrawl) [inline packages/rest/__tests__/DiscordAPIError.test.ts:68]: We should just let prettier do its thing here.

File: packages/rest/__tests__/DiscordAPIError.test.ts [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+		50035,
+		400,
+		'PATCH',
+		'https://discord.com/api/v8/guilds/:id',
+	);
+
+	expect(error.code).toBe(50035);
+	expect(error.message).toBe(
+		[
+			'Invalid Form Body', //
```
Comment (iCrawl) [inline packages/rest/__tests__/DiscordAPIError.test.ts:98]: We should just let prettier do its thing here.

File: packages/rest/src/lib/CDN.ts [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	/**
+	 * Generates a user avatar URL for a user's avatar.
+	 * @param userID The user ID that has the icon
+	 * @param avatarHash The hash provided by Discord for this avatar
+	 * @param options Optional options for the avatar
+	 */
+	public userAvatar(userID: string, avatarHash: string, { dynamic = false, ...options }: ImageURLOptions = {}): string {
+		if (dynamic && avatarHash.startsWith('a_')) options.extension = 'gif';
```
Comment (iCrawl) [inline packages/rest/src/lib/CDN.ts:119]: I would propose for consistency's sake to expand all if statements to be easier on the diff and potential future additions.

```suggestion
		if (dynamic && avatarHash.startsWith('a_')) {
			options.extension = 'gif';
		}
```

File: packages/rest/src/lib/CDN.ts [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+	 * @param options The extension/size options for the link
+	 */
+	private makeURL(base: string, { extension = 'png', size }: ImageURLOptions = {}): string {
+		extension = String(extension).toLowerCase() as ImageExtension;
+
+		if (!ALLOWED_EXTENSIONS.includes(extension))
+			throw new RangeError(
+				`Invalid extension provided: ${extension}\nMust be one of: ${ALLOWED_EXTENSIONS.join(', ')}`,
+			);
+		if (size && !ALLOWED_SIZES.includes(size))
```
Comment (iCrawl) [inline packages/rest/src/lib/CDN.ts:135]:
```suggestion
		if (size && !ALLOWED_SIZES.includes(size)) {
			throw new RangeError(`Invalid size provided: ${size}\nMust be one of: ${ALLOWED_SIZES.join(', ')}`);
		}
```

File: packages/rest/src/lib/CDN.ts [hunk 33]
```diff
@@ -0,0 +1,143 @@
...
+			throw new RangeError(`Invalid size provided: ${size}\nMust be one of: ${ALLOWED_SIZES.join(', ')}`);
+
+		const url = new URL(`${this.base}${base}.${extension}`);
+		if (size) url.searchParams.set('size', String(size));
```
Comment (iCrawl) [inline packages/rest/src/lib/CDN.ts:139]:
```suggestion
		if (size) {
			url.searchParams.set('size', String(size));
		}
```

File: packages/rest/src/lib/RequestManager.ts [hunk 34]
```diff
@@ -0,0 +1,292 @@
...
+	 * Formats the request data to a usable format for fetch
+	 * @param request The request data
+	 */
+	private resolveRequest(request: InternalRequest): { url: string; fetchOptions: RequestInit } {
+		const { options } = this;
+
+		let query = '';
+
+		// If a query option is passed, use it
+		if (request.query) query = `?${request.query.toString() as string}`;
```
Comment (iCrawl) [inline packages/rest/src/lib/RequestManager.ts:194]:
```suggestion
		if (request.query) {
			query = `?${request.query.toString() as string}`;
		}
```

File: packages/rest/src/lib/RequestManager.ts [hunk 35]
```diff
@@ -0,0 +1,292 @@
...
+
+		// Create the required headers
+		const headers: RequestHeaders = {
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth !== false) {
+			// If we haven't received a token, throw an error
+			if (!this.#token) throw new Error('Expected token to be set for this request, but none was present');
```
Comment (iCrawl) [inline packages/rest/src/lib/RequestManager.ts:204]:
```suggestion
			if (!this.#token) {
				throw new Error('Expected token to be set for this request, but none was present');
			}
```

File: packages/rest/src/lib/RequestManager.ts [hunk 36]
```diff
@@ -0,0 +1,292 @@
...
+			const formData = new FormData();
+
+			// Attach all files to the request
+			for (const attachment of request.attachments) {
+				formData.append(attachment.fileName, attachment.rawBuffer, attachment.fileName);
+			}
+
+			// If a JSON body was added as well, attach it to the form data
+			// eslint-disable-next-line no-eq-null
+			if (request.body != null) formData.append('payload_json', JSON.stringify(request.body));
```
Comment (iCrawl) [inline packages/rest/src/lib/RequestManager.ts:232]:
```suggestion
			if (request.body != null) {
				formData.append('payload_json', JSON.stringify(request.body));
			}
```

File: packages/rest/src/lib/RequestManager.ts [hunk 37]
```diff
@@ -0,0 +1,292 @@
...
+			.replace(/\/reactions\/(.*)/, '/reactions/:reaction');
+
+		let exceptions = '';
+
+		// Hard-Code Old Message Deletion Exception (2 week+ old messages are a different bucket)
+		// https://github.com/discord/discord-api-docs/issues/1295
+		if (method === RequestMethod.Delete && baseRoute === '/channels/:id/messages/:id') {
+			const id = /\d{16,19}$/.exec(endpoint)![0];
+			const snowflake = DiscordSnowflake.deconstruct(id);
+			if (Date.now() - Number(snowflake.timestamp) > 1000 * 60 * 60 * 24 * 14) exceptions += '/Delete Old Message';
```
Comment (iCrawl) [inline packages/rest/src/lib/RequestManager.ts:283]:
```suggestion
			if (Date.now() - Number(snowflake.timestamp) > 1000 * 60 * 60 * 24 * 14) {
				exceptions += '/Delete Old Message';
			}
```

File: packages/rest/src/lib/errors/DiscordAPIError.ts [hunk 38]
```diff
@@ -0,0 +1,82 @@
...
+			return yield `${key.length ? `${key}[${obj.code}]` : `${obj.code}`}: ${obj.message}`.trim();
+		}
+
+		for (const [k, v] of Object.entries(obj)) {
+			const nextKey = k.startsWith('_') ? key : key ? (Number.isNaN(Number(k)) ? `${key}.${k}` : `${key}[${k}]`) : k;
+
+			if (typeof v === 'string') {
+				yield v;
+			} else if (isErrorGroupWrapper(v)) {
+				for (const error of v._errors) yield* this.flattenDiscordError(error, nextKey);
```
Comment (iCrawl) [inline packages/rest/src/lib/errors/DiscordAPIError.ts:76]:
```suggestion
				for (const error of v._errors) {
					yield* this.flattenDiscordError(error, nextKey);
				}
```

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 39]
```diff
@@ -0,0 +1,221 @@
...
+	private async runRequest(routeID: RouteData, url: string, options: RequestInit, retries = 0): Promise<unknown> {
+		const controller = new AbortController();
+		const timeout = setTimeout(() => controller.abort(), this.manager.options.timeout);
+		let res: Response;
+
+		try {
+			res = await fetch(url, { ...options, signal: controller.signal });
+		} catch (error) {
+			// Retry the specified number of times for possible timed out requests
+			if (error.name === 'AbortError' && retries !== this.manager.options.retries)
```
Comment (iCrawl) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:134]:
```suggestion
			if (error.name === 'AbortError' && retries !== this.manager.options.retries) {
				return this.runRequest(routeID, url, options, ++retries);
			}
```

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 40]
```diff
@@ -0,0 +1,221 @@
...
+
+		// Amount of time in milliseconds until we should retry if rate limited (globally or otherwise)
+		if (retry) retryAfter = Number(retry) * 1000 + this.manager.options.offset;
+
+		// Handle buckets via the hash header retroactively
+		if (hash && hash !== this.hash) {
+			// Let library users know when rate limit buckets have been updated
+			this.debug(
+				[
+					'Received bucket hash update', //
```
Comment (iCrawl) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:165]: We should just let prettier do its thing here.

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 41]
```diff
@@ -0,0 +1,221 @@
...
+					`  Retry After    : ${retryAfter}ms`,
+				].join('\n'),
+			);
+			// Wait the retryAfter amount of time before retrying the request
+			await sleep(retryAfter);
+			// Since this is not a server side issue, the next request should pass, so we don't bump the retries counter
+			return this.runRequest(routeID, url, options, retries);
+		} else if (res.status >= 500 && res.status < 600) {
+			// Retry the specified number of times for possible server side issues
+			if (retries !== this.manager.options.retries) return this.runRequest(routeID, url, options, ++retries);
```
Comment (iCrawl) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:203]:
```suggestion
			if (retries !== this.manager.options.retries) {
				return this.runRequest(routeID, url, options, ++retries);
			}
```

File: packages/rest/src/lib/utils/utils.ts [hunk 42]
```diff
@@ -0,0 +1,10 @@
+import type { Response } from 'node-fetch';
+
+/**
+ * Converts the response to usable data
+ * @param res The node-fetch response
+ */
+export function parseResponse(res: Response): Promise<unknown> {
+	if (res.headers.get('Content-Type')?.startsWith('application/json')) return res.json();
```
Comment (iCrawl) [inline packages/rest/src/lib/utils/utils.ts:8]:
```suggestion
	if (res.headers.get('Content-Type')?.startsWith('application/json')) {
		return res.json();
	}
```

**Review by kyranet — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]
